### PR TITLE
feat(ModerationRequest): Added pagination for Open/Closed Moderation. Optimized loading of Moderation Request

### PR DESCRIPTION
**Pull request by JaideepPalit**
> Added pagination for Open/Closed Moderation. Optimized loading of Moderation Request.
> * Which issue is this pull request belonging to and how is it solving it? (*#1111*)
> * Did you add or update any new dependencies that are required for your change? - No

> Have you implemented any additional tests? - No

### Checklist
Must:
- [X] All related issues are referenced in commit messages and in PR

Signed-off-by: Jaideep Palit <jaideep.palit@siemens.com>

### CODE REVIEW

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/moderation/view.jsp [hunk 1]
```diff
@@ -363,71 +380,60 @@ AUI().use('liferay-portlet-url', function () {
...
+            return "";
+        }
+
+        function detailUrl(name, type, row) {
+            let url = makeModerationViewUrl(row.id);
+            let viewUrl = $("<a></a>").attr("href",url).css("word-break","break-word").text(name);
+            return viewUrl[0].outerHTML;
+        }
+
+        var homeUrl = themeDisplay.getURLHome().replace(/\/web\//, '/group/');
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/resources/META-INF/resources/html/moderation/view.jsp:432]: please refer `link.js` for url formation

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/moderation/ModerationPortlet.java [hunk 2]
```diff
@@ -803,4 +839,104 @@ private String getDefaultLicenseInfoHeaderText() {
...
+            Collections.sort(modList, PortletUtils.compareByModerationState(isAsc));
+            break;
+        default:
+            break;
+        }
+
+        return modList;
+    }
+
+    private String displayUserCollection(Set<String> values) {
```
Comment (smrutis1) [inline frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/moderation/ModerationPortlet.java:914]: Is it possible to handle this in UI like using tag `<sw360:DisplayUserEmailCollection>` and `<sw360:DisplayEnum>` ?

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java [hunk 3]
```diff
@@ -289,13 +290,20 @@ public int getCountByProjectId(String id) {
...
     @NotNull
     private Set<Project> filterAccessibleProjectsByIds(User user, Set<String> searchIds) {
-        final Set<Project> accessibleProjects = getAccessibleProjects(user);
-
         final Set<Project> output = new HashSet<>();
-        for (Project accessibleProject : accessibleProjects) {
-            if (searchIds.contains(accessibleProject.getId())) output.add(accessibleProject);
-        }
 
+        searchIds.stream().filter(CommonUtils::isNotNullEmptyOrWhitespace).forEach(searchId -> {
```
Comment (smrutis1) [inline backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java:295]: I hope, this is not related to `moderation request`?

**Review by smrutis1 — CHANGES_REQUESTED** [review]
some comments

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java (hunk 3 above)
Comment (JaideepPalit) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java:295]: `filterAccessibleProjectsByIds(User user, Set<String> searchIds)` is getting called when a moderation request is opened.
Called from ModerationPortlet.class -> `renderProjectModeration()` ->`prepareProject()`.
Similarly called from ModerationPortlet.class -> `renderComponentModeration()` ->`prepareComponent()`.
Similarly called from ModerationPortlet.class -> `renderReleaseModeration()` ->`prepareRelease(`).

Therefore optimized code here, since it was loading all projects before.

File: frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/moderation/ModerationPortlet.java (hunk 2 above)
Comment (JaideepPalit) [reply frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/moderation/ModerationPortlet.java:914]: Not sure whether it is possible to handle this in UI using tags <sw360:DisplayUserEmailCollection> and <sw360:DisplayEnum>.
Since tags are part of jsp and converted once on page loading. Moderation Datatable is ajax and js based.
However I tried it, but without positive outcome.
Please do suggest if you have a solution

File: frontend/sw360-portlet/src/main/resources/META-INF/resources/html/moderation/view.jsp (hunk 1 above)
Comment (JaideepPalit) [reply frontend/sw360-portlet/src/main/resources/META-INF/resources/html/moderation/view.jsp]: done

**Comment by JaideepPalit** [comment]
Thanks for review and test. Implemented all possible review comments.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java (hunk 3 above)
Comment (smrutis1) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ProjectRepository.java:295]: Missed it. Thanks for the explanation.

**Comment by smrutis1** [comment]
Tested in my local machine with limited datasets. Please let me know if I am missing something.

1) One datatable warning keeps popping up during page reload.
image
2) If the table is empty there is no empty message like `No moderation request found`

image

**Comment by JaideepPalit** [comment]
Thanks for test. 
From screenshots , it indicates issue in closed Moderation request table when no data present.
- Fixed

**Review by smrutis1 — APPROVED** [review]
Code looks good to me. Tested the reported issue and basics of moderation handling.

**Comment by mcjaeger** [comment]
stage instance with branch: fix/ExcelExport (eefcf17) loads in 41secs and this branch in 54 secs ..

**Comment by mcjaeger** [comment]
retest after new deployment, actually, new code seem to be slower than old code on stage instance.
